File: 백준/Gold/20437. 문자열 게임 2/문자열 게임 2.java
import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int t = Integer.parseInt(br.readLine());
        while(t-->0){
            String w = br.readLine().trim();
            int k = Integer.parseInt(br.readLine());

            Map<Character,Integer> cntMap = new HashMap<>();
            Map<Character, List<Integer>> indexMap = new HashMap<>();

            for(int i=0;i<w.length();i++){
                char c= w.charAt(i);
                cntMap.put(c,cntMap.getOrDefault(c,0)+1);
                if(!indexMap.containsKey(c)){
                    indexMap.put(c, new ArrayList<>());
                }
                indexMap.get(c).add(i);
            }
            boolean check = false;
            int minAns = Integer.MAX_VALUE;
            int maxAns = -1;

            for(char key:indexMap.keySet()){
                if(cntMap.get(key)<k)continue;
                check = true;
                List<Integer> indexs = indexMap.get(key);
                for(int i=0;i<=indexs.size()-k;i++){
                    int length = indexs.get(i+k-1)-indexs.get(i)+1;
                    minAns = Math.min(minAns,length);
                    maxAns = Math.max(maxAns,length);
                }
            }
            if(check){
                System.out.println(minAns+" "+maxAns);
            }else{
                System.out.println(-1);
            }
        }
    }
}

File: 백준/Gold/20437. 문자열 게임 2/문자열 게임 2.py
import sys
input = sys.stdin.readline

t = int(input())
for _ in range(t):
    w = input().rstrip()
    k = int(input())
    cnt_dict = {}

    # 각 문자 개수 세기
    for char in w:
        cnt_dict[char] = cnt_dict.get(char, 0) + 1

    check = False
    max_ans = -1
    min_ans = float('inf')  # 큰 값으로 초기화

    check_dict = {}

    # k번 이상 등장하는 문자들의 인덱스 저장
    for i in range(len(w)):
        if cnt_dict[w[i]] < k:
            continue
        check = True
        if w[i] not in check_dict:
            check_dict[w[i]] = []
        check_dict[w[i]].append(i)

    #최소, 최대 길이 계산
    for key, v in check_dict.items():
        if len(v) >= k:
            for i in range(len(v) - k + 1):
                length = v[i + k - 1] - v[i] + 1
                max_ans = max(max_ans, length)
                min_ans = min(min_ans, length)

    if check:
        print(min_ans, max_ans)
    else:
        print(-1)
